### PR TITLE
BaSP-TG-09: delete and update a super admin

### PR DESCRIPTION
**Pull request by Maxi-SR**
Feature to delete a super admin by id and update it too.

### CODE REVIEW

File: src/models/SuperAdmins.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+const superAdminSchema = new Schema({
+  name: { type: String, required: true },
+  lastName: { type: String, required: true },
+  email: { type: String, required: true },
+  password: { type: String, required: true },
+});
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:9]: A lineBreak is necessary here in between line 9 and 10

File: src/models/SuperAdmins.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+const superAdminSchema = new Schema({
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:4]: A lineBreak is necessary here in between line 3 and 4

File: src/routes/superAdmins.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import express from 'express';
+import {
+  getAllSuperAdmins, getSuperAdminById, createSuperAdmin, deleteSuperAdmin, updateSuperAdmin,
+} from '../controllers/superAdmins';
+import { validateCreation, validateEdit } from '../validations/superAdmins';
+
+const router = express.Router();
+router.get('/', getAllSuperAdmins);
+router.get('/:id', getSuperAdminById);
+router.post('/', validateCreation, createSuperAdmin);
```
Comment (martinraveglia) [inline src/routes/superAdmins.js:10]: Why are these methods here, if this PR is of delete and update features?

File: src/controllers/superAdmins.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  } catch (error) {
+    return res.status(404).json({
+      message: error.toString(),
+      data: undefined,
+      error: true,
+    });
+  }
+};
+
+const createSuperAdmin = async (req, res) => {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:50]: Why is this method here, if this PR is of delete and update features?

File: src/controllers/superAdmins.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return res.status(statusCode).json({
+      message: error.toString(),
+      data: undefined,
+      error: true,
+    });
+  }
+};
+
+const getSuperAdminById = async (req, res) => {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:29]: Why is this method here, if this PR is of delete and update features?

File: src/controllers/superAdmins.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
+import SuperAdmins from '../models/SuperAdmins';
+
+const getAllSuperAdmins = async (req, res) => {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:3]: Why is this method here, if this PR is of delete and update features?

File: src/controllers/superAdmins.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  } catch (error) {
+    return res.status(404).json({
+      message: error.toString(),
+      error: true,
+      data: undefined,
+    });
+  }
+};
+
+const deleteSuperAdmin = async (req, res) => {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:74]: We have been using async await on every method, let's update this one for consistency purposes

File: src/controllers/superAdmins.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return res.status(204).json({
+      message: `Super admin with id: ${idSuperAdmin} deleted`,
+      data: superAdmin,
+      error: false,
+    });
+  });
+};
+
+const updateSuperAdmin = (req, res) => {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:92]: We have been using async await on every method, let's update this one for consistency purposes

File: src/validations/superAdmins.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  }
+  return next();
+};
+
+const validateEdit = (req, res, next) => {
+  const pattern = /^[^@]+@[^@]+\.[a-zA-Z]{2,}$/;
+  const superAdminValidation = Joi.object({
+    name: Joi.string().min(3).max(50),
+    lastName: Joi.string().min(3).max(50),
+    email: Joi.string().pattern(new RegExp(pattern)),
```
Comment (martinraveglia) [inline src/validations/superAdmins.js:27]: Joi has a built-in method to validate emails, let's use that one

File: src/validations/superAdmins.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    password: Joi.string().alphanum().min(8),
+  });
+  const validation = superAdminValidation.validateAsync(req.body, { abortEarly: false });
+  if (validation.error) {
+    return res.status(400).json({
+      message: `There was an error: ${validation.error.details[0].message}`,
```
Comment (martinraveglia) [inline src/validations/superAdmins.js:33]: if we have more than one validation error (in more than one field), all errors should be sent

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
-import routes from './routes';
+import superAdminRoute from './routes/superAdmins';
 
 const app = express();
 const port = process.env.PORT || 3000;
 const MONGO_URL = 'mongodb+srv://RadiumA:RadiumA@trackgenix.r6u6do6.mongodb.net/?retryWrites=true&w=majority';
 
 app.use(express.json());
-app.use(routes);
+app.use('/superAdmins', superAdminRoute);
```
Comment (martinraveglia) [inline src/index.js:10]: This belongs to the routes index file!!

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/routes/superAdmins.js (hunk 3 above)
Comment (Maxi-SR) [reply src/routes/superAdmins.js:10]: I delete it after check that delete and update works

File: src/index.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import express from 'express';
 import mongoose from 'mongoose';
-import routes from './routes';
+import routes from './routes/index';
```
Comment (martinraveglia) [inline src/index.js:3]:
```suggestion
import routes from './routes';
```

**Review by martinraveglia — CHANGES_REQUESTED** [review]

**Review by martinraveglia — APPROVED** [review]

File: src/controllers/superAdmins.js [hunk 13]
```diff
@@ -1,5 +1,81 @@
...
+      throw new Error('Super admin not found');
+    }
+    return res.status(200).json({
+      message: `Super admin with id: ${idSuperAdmin} found`,
+      data: superAdmin,
+      error: false,
+    });
+  } catch (error) {
+    let statusCode = 400;
+    if (error.message.includes('Super admin not found')) {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:44]: This will never match the error

**Review by martinraveglia — CHANGES_REQUESTED** [review]

**Review by martinraveglia — APPROVED** [review]

**Pull request merged**